### PR TITLE
Enhancment to support HostedClusterInfrastructure CRD

### PR DESCRIPTION
**Pull request by jnpacker**
Signed-off-by: Joshua Packer <jpacker@redhat.com>

### CODE REVIEW

**Comment by jnpacker** [comment]
So I've mocked up a hostedclusterinfrastructure_type.go file.

Basic form is:
```yaml
apiVersion: hypershift.openshift.io/v1alpha1
kind: HostedClusterInfrastructure
metadata:
  name: my-infra-id
spec:
  platform:
    type: AWS
  dns:
    baseDomain: dev06.red-chesterfield.com
  cloudProvider:
    name: my-provider-secret
```

The `oc get hci` looks like this:
```script
NAME    TYPE   INFRA   IAM   PROVIDER REF   PHASE
test1   AWS
```

Should I make a pull toward the `openshift`/`hypershift` repo?

**Comment by jnpacker** [comment]
FYI @csrwng

File: enhancements/hypershift/hostedClusterInfrastructure.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+### Goals
+- Provide a consumable API for HostedClusterInfrastructure that satisfies existing ACM/MCE use cases and expected user stories(flows)
+
+### Non-Goals
+- The consumer usecase also compliments the cluster create templating flow
+
+## Proposal
+### API
+[HostedClusterInfrastructure](https://github.com/openshift/hypershift/blob/27c0a432bdc8d702f1cdb2a2f3f25e5ae6fbee7d/api/v1alpha1/hostedclusterinfrastructure_types.go) is the consumer facing API exposed for Cloud Infrastructure creation and destroy.
```
Comment (enxebre) [inline enhancements/hypershift/hostedClusterInfrastructure.md:66]: broken link?

File: enhancements/hypershift/hostedClusterInfrastructure.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+## Glossary
+- Untangling HypershiftDeployment, HostedCluster and NodePool Google document - [Google Document - Option D](https://docs.google.com/document/d/1VabhUQa_uZlWO6mf9UHS5lGtnH_X915cih4ARJbNibc/edit#heading=h.iuee7tji8zz4)
+
+## Motivation
+ACM and MCE primary Cloud Provider use cases already provision these resources and manage these resources, this provides feature parity with Standalone flows that ACM Cluster Administrator's are used to.  This also allows to additional opporunities:
+  1. Infrastructure can be pre-created using an ACM Cloud Provider credential or by pre-populating the Custom Resource with values and applying it to the ACM/MCE hub. These can then be consumed by the ACM/MCE UI or CLI, and is a nice option, as the AWS credential does not need to be exposed to the user creating the HostedCluster.
+  2. You can re-use the HostedClusterInfrastructure resource between HostedClusters. This is especially useful in Azure, where it takes longer to create infrastructure then the hosted control plane.
```
Comment (enxebre) [inline enhancements/hypershift/hostedClusterInfrastructure.md:51]: do we know of uses cases where sharing infra in azure is desired? can we elaborate them?

File: enhancements/hypershift/hostedClusterInfrastructure.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+N/A.
+### Version Skew Strategy
+N/A.
+### Operational Aspects of API Extensions
+N/A.
+#### Failure Modes
+#### Support Procedures
+
+## Alternatives
+
```
Comment (enxebre) [inline enhancements/hypershift/hostedClusterInfrastructure.md:132]: Can we elaborate this section?

File: enhancements/hypershift/hostedClusterInfrastructure.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+      - The resource just contains user provided values that can be consumed by the UI or CLI
+- CLI will be able to create or render the HostedClusterInfrastructure resource
+- CLI will be able to consume the HostedClusterInfrastructure resource
+- User may create/apply the resource with all values
+- Delete HostedClusterInfrastructure the controller will reconcile
+  - First it will determine if a Cloud Provider credential is present
+    - If Present:
+      - It will delete all the infrastructure resources found in the cr before allowing the cr to be removed
+    - If Not Present
+      - The cr will just be deleted
```
Comment (davidvossel) [inline enhancements/hypershift/hostedClusterInfrastructure.md:87]: in a situation where the credential existed at one point, but was deleted before the HCI, wouldn't this kind of "silently" leave infra orphaned? If so, should we account for this?

File: enhancements/hypershift/hostedClusterInfrastructure.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+- Consume an existing HostedClusterInfrastructure when creating a HostedCluster and NodePool resource
+- If connected to the Hosting Cluster(Management Cluster) When creating infrastructure apply the HCI cr as part of the creation process
+
+#### Console
+- Allow creation of infrastructure via the HCI cr
+- Allow deletion of infrastructure via the HCI cr
+- Allow Admin to pre-populate infrastructure for use by a consumer who just creates the HostedCluster and NodePool cr's
+- (spike) Report status of resources in HCI
+### Workflow Description
+The [HostedClusterInfrastructure API](https://github.com/openshift/hypershift/blob/27c0a432bdc8d702f1cdb2a2f3f25e5ae6fbee7d/api/v1alpha1/hostedclusterinfrastructure.go)
```
Comment (davidvossel) [inline enhancements/hypershift/hostedClusterInfrastructure.md:102]: 404

File: enhancements/hypershift/hostedClusterInfrastructure.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+  1. Infrastructure can be pre-created using an ACM Cloud Provider credential or by pre-populating the Custom Resource with values and applying it to the ACM/MCE hub. These can then be consumed by the ACM/MCE UI or CLI, and is a nice option, as the AWS credential does not need to be exposed to the user creating the HostedCluster.
+  2. You can re-use the HostedClusterInfrastructure resource between HostedClusters. This is especially useful in Azure, where it takes longer to create infrastructure then the hosted control plane.
+
+### User Stories
+- As a Cluster Administrator I want to create one or more HostedClusterInfrastructure resources for my team to consume
+- As a Cluster Administrator, I want to choose an ACM Cloud Provider credential and deploy a Cloud Provider HostedCluster from start to finish
+- As a Cluster Consumer, I want to choose an available HostedClusterInfrastructure resource and have a HostedCluster deploy, without knowing or having to provide an AWS credential
+
+### Goals
```
Comment (davidvossel) [inline enhancements/hypershift/hostedClusterInfrastructure.md:58]: would we consider consistency across the cli and web console also a goal?  I think explicitly calling this out helps narrow down the discussion when it comes to alternatives.

File: enhancements/hypershift/hostedClusterInfrastructure.md (hunk 2 above)
Comment (jnpacker) [reply enhancements/hypershift/hostedClusterInfrastructure.md:51]: Added. It was to either pre-create to improve speed, trade off is cost of the resources.  Re-use in dev an pipeline cases were also mentioned in the untaggle doc. Pre-creating does save 5-8min off create of the cluster.

File: enhancements/hypershift/hostedClusterInfrastructure.md (hunk 6 above)
Comment (jnpacker) [reply enhancements/hypershift/hostedClusterInfrastructure.md:71]: agreed

File: enhancements/hypershift/hostedClusterInfrastructure.md (hunk 4 above)
Comment (jnpacker) [reply enhancements/hypershift/hostedClusterInfrastructure.md:101]: Yes, in this scenario, there is a status check for the cloud provider (borrowing how it looked in HypershiftDeployment), and in the status for the HostedClusterInfrastructure, we want to differentiate whether the resource was applied successfully or just provided without ever being applied.  We would detect this case an show an appropriate error and have a finalizer blocking deletion of the HostedClusterResource.

File: enhancements/hypershift/hostedClusterInfrastructure.md (hunk 1 above)
Comment (jnpacker) [reply enhancements/hypershift/hostedClusterInfrastructure.md]: fixed

File: enhancements/hypershift/hostedClusterInfrastructure.md (hunk 5 above)
Comment (jnpacker) [reply enhancements/hypershift/hostedClusterInfrastructure.md]: fixed.

File: enhancements/hypershift/hostedClusterInfrastructure.md (hunk 3 above)
Comment (jnpacker) [reply enhancements/hypershift/hostedClusterInfrastructure.md:148]: Added some of the alternatives we discussed in Untangled.

File: enhancements/hypershift/hostedClusterInfrastructure.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+## Glossary
+- Untangling HypershiftDeployment, HostedCluster and NodePool Google document - [Google Document - Option D](https://docs.google.com/document/d/1VabhUQa_uZlWO6mf9UHS5lGtnH_X915cih4ARJbNibc/edit#heading=h.iuee7tji8zz4)
+
+## Motivation
+ACM and MCE primary Cloud Provider use cases already provision these resources and manage these resources, this provides feature parity with Standalone flows that ACM Cluster Administrator's are used to.  This also allows to additional opporunities:
+  1. Infrastructure can be pre-created using an ACM Cloud Provider credential or by pre-populating the Custom Resource with values and applying it to the ACM/MCE hub. These can then be consumed by the ACM/MCE UI or CLI, and is a nice option, as the AWS credential does not need to be exposed to the user creating the HostedCluster.
+  2. You can re-use the HostedClusterInfrastructure resource between HostedClusters. This is especially useful in Azure, where it takes longer to create infrastructure then the hosted control plane.
+
+### User Stories
+- As a Cluster Administrator I want to create one or more HostedClusterInfrastructure resources for my team to consume
```
[the remark this answers is not in the document]
Comment (jnpacker) [reply enhancements/hypershift/hostedClusterInfrastructure.md]: It's not explicitly called out as this is why its better, but most of the user stories are not possible without create infra.

Example: Pre populating infrastructure so the UI can pick from it, GitOps flow to create or store infrastructure data that can be consumed by Console, CLI and HostedCluster (assuming we make it HCI aware or populate it with the HCI controller)

**Comment by jnpacker** [comment]
@enxebre @davidvossel I'm interested in your thoughts and opinions on the non-referenced resources.
ie. There are at least two subnets, there are EIPs and DHCP options.  

In my initial types.go mock up I did not include these, but I think all trackable resources ID's should probably be present.
I was thinking some of these more intermediate might be just shown in the status when created or detected. As they have bearing on the configuration, but are not required to be passed to HC or NP.

**Comment by jnpacker** [comment]
nudge: @enxebre @davidvossel

**Comment by davidvossel** [comment]
>I'm interested in your thoughts and opinions on the non-referenced resources.
ie. There are at least two subnets, there are EIPs and DHCP options.
> In my initial types.go mock up I did not include these, but I think all trackable resources ID's should probably be present.
I was thinking some of these more intermediate might be just shown in the status when created or detected. As they have bearing on the configuration, but are not required to be passed to HC or NP. 

When I originally looked at this the example types.go files returned a 404, so I'm just seeing those now.

The separation between what exists in the HCI's Spec and Status is different than what I expected. I thought the spec would just be user input (things like "region", "aws creds", etc...) and that the status would the be aws ids associated with whatever the HCI provisioned.

Using the hypershift cli as an analogy. The HCI Spec would have fields similar to the args passed in to the cli (minus the output-file). And the HCI Status would essentially contain the data in output-file, plus conditions indicating the reconcile status.

**cli example for reference**
```
hypershift create infra aws --name CLUSTER_NAME \
    --aws-creds AWS_CREDENTIALS_FILE \
    --base-domain BASEDOMAIN \
    --infra-id INFRA_ID \
    --region REGION \
    --output-file OUTPUT_INFRA_FILE
```

Instead, it looks like the spec has both the user input and output combined (maybe?) and that the status is mostly conditions. I had been working with the mental model that the spec would be completely isolated to user input, and all the tracking of cloud infra resources would be written to the status field.

So, to answer your question. In the model I had in my head for this api, I would at least expect to see references in the Status section for all the infra parts that need to be feed into the HostedCluster/NodePool. Tracking all the created infra pieces on the HCI's status could have value for visibility during debugging.

**Comment by jnpacker** [comment]
@davidvossel That is what I originally looked at doing, but we had written in untangle the use case to allow the user to provide the values themselves.

Though I'm thinking now, that if a user wants to provide some values themselves, we require them to TAG them correctly, and we would just pick them up and include the ID values in the status.  This way HCI and CLI would work the same (if we don't move CLI to only use HCI)

**Comment by jnpacker** [comment]
@davidvossel @enxebre here is the commit incorporating the original status. https://github.com/jnpacker/hypershift/commit/300958f60f1cd30cabb1c25a4bf8acdb13708836

Both create infra and create iam are implemented.

**Comment by jnpacker** [comment]
FYI @rokej

**Comment by jnpacker** [comment]
Here is the Custom Resource example after running create infra and create iam
```yaml
apiVersion: hypershift.openshift.io/v1alpha1
kind: HostedClusterInfrastructure
metadata:
  name: test01
  namespace: clusters
spec:
  altInfraID: test01
  baseDomain: XYZ....com
  cloudProvider: {}
  platform:
    aws:
      region: us-west-1
    type: AWS
status:
  associatedInfrastructure:
    dns:
      localZoneID: XYZ...
    platform:
      aws:
        oidcProviderARN: arn:aws:iam::XYZ...:oidc-provider/XYZ....amazonaws.com/test01
  coreInfrastructure:
    dns:
      baseDomain: XYZ...com
      privateZoneID: XYZ...
      publicZoneID: XYZ...
    issuerURL: https://XYZ.../test01
    networking:
      machineNetwork:
      - cidr: 10.0.0.0/16
    platform:
      aws:
        instanceProfile: test01-worker
        rolesRef:
          controlPlaneOperatorARN: arn:aws:iam::XYZ...:role/test01-control-plane-operator
          imageRegistryARN: arn:aws:iam::XYZ...:role/test01-openshift-image-registry
          ingressARN: arn:aws:iam::XYZ...:role/test01-openshift-ingress
          kubeCloudControllerARN: arn:aws:iam::XYZ...:role/test01-cloud-controller
          networkARN: arn:aws:iam::XYZ...:role/test01-cloud-network-config-controller
          nodePoolManagementARN: arn:aws:iam::XYZ...:role/test01-node-pool
          storageARN: arn:aws:iam::XYZ...:role/test01-aws-ebs-csi-driver-controller
        securityGroups:
        - id: sg-XYZ...
        vpc: vpc-XYZ...
        zones:
        - subnet:
            id: subnet-XYZ...
          zone: us-west-1a
  phase: READY

**Comment by davidvossel** [comment]
> Though I'm thinking now, that if a user wants to provide some values themselves, we require them to TAG them correctly, and we would just pick them up and include the ID values in the status. This way HCI and CLI would work the same (if we don't move CLI to only use HCI)

yep, that sounds reasonable way to satisfy both requirements and keep the API consistent.

**Comment by jnpacker** [comment]
Made `HostedClusterInfrastructure` PR https://github.com/openshift/hypershift/pull/1793

File: enhancements/hypershift/hostedClusterInfrastructure.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+      - It will delete all the infrastructure resources found in the cr before allowing the cr to be removed
+    - If Not Present
+      - The cr will just be deleted
+
+### CLI
+Supported CLI capabilities:
+- Create a HostedClusterInfrastructure using an AWS/Azure credential or secret
+- Render a HostedClusterInfrastructure using an AWS/Azure credential or secret (for Git or manual create/apply)
+- Consume an existing HostedClusterInfrastructure when creating a HostedCluster and NodePool resource
+- If connected to the Hosting Cluster(Management Cluster) When creating infrastructure apply the HCI cr as part of the creation process
```
Comment (enxebre) [inline enhancements/hypershift/hostedClusterInfrastructure.md:117]: typo cr->CR/custom resource

File: enhancements/hypershift/hostedClusterInfrastructure.md [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+  The result HostedClusterInfrastructure object can be given get/list verbs to the consumer who wants to create a cluster. 
+  This consumer would then only need CRUD for HostedCluster and NodePool and be able to use the console to pick a HostedClusterInfrastructure resource and then create a HostedCluster.  No AWS credential needs to be passed to the consumer.  
+  * The second is that for slow cloud providers like Azure, the infrastructure can be pre-created to improve cluster deployment time (in dev, test, and pipeline scenarios). 
+  It also seems possible to re-use the infrastructure between clusters, so treating HostedClusterInfrastructure as an entity outside the hostedCluster promotes scenarios of re-use (saves time, especially in Azure. Most likely a dev or pipeline user case).
+  * The third is that the HostedClusterInfrastructure YAML can be committed to Git and applied using ArgoCD, Flux or ACM Subscriptions to create a gitOps flow for infrastructure creation and deletion.
+  * The fourth is to apply HostedClusterInfrastructure with value already present and either include or not include the cloud provider credential.
```
Comment (enxebre) [inline enhancements/hypershift/hostedClusterInfrastructure.md:63]: >* The fourth is to apply HostedClusterInfrastructure with value already present and either include or not include the cloud provider credential.

to clarify Is this to "adopt" existing infra?

File: enhancements/hypershift/hostedClusterInfrastructure.md [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+  1. Infrastructure can be pre-created using an ACM Cloud Provider credential or by pre-populating the Custom Resource with values and applying it to the ACM/MCE hub. These can then be consumed by the ACM/MCE UI or CLI, and is a nice option, as the AWS credential does not need to be exposed to the user creating the HostedCluster.
+  2. You can re-use the HostedClusterInfrastructure resource between HostedClusters. This is especially useful in Azure, where it takes longer to create infrastructure then the hosted control plane.
+
+### User Stories
+- As a Cluster Administrator I want to create one or more HostedClusterInfrastructure resources for my team to consume.
+  
+  This approach pre-creates the HostedClusterInfrastructure before the consumer creates the HostedCluster and NodePool resources. This approach has two advantage for ACM use cases:
+  * The first advantage is that and Administrator can create the infrastructure resources using either their account or a service account stored in the ACM Cloud Provider credential (AWS or Azure cred).
```
Comment (enxebre) [inline enhancements/hypershift/hostedClusterInfrastructure.md:57]: typo and/an

File: enhancements/hypershift/hostedClusterInfrastructure.md [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+
+## Motivation
+ACM and MCE primary Cloud Provider use cases already provision these resources and manage these resources, this provides feature parity with Standalone flows that ACM Cluster Administrator's are used to.  This also allows to additional opporunities:
+  1. Infrastructure can be pre-created using an ACM Cloud Provider credential or by pre-populating the Custom Resource with values and applying it to the ACM/MCE hub. These can then be consumed by the ACM/MCE UI or CLI, and is a nice option, as the AWS credential does not need to be exposed to the user creating the HostedCluster.
+  2. You can re-use the HostedClusterInfrastructure resource between HostedClusters. This is especially useful in Azure, where it takes longer to create infrastructure then the hosted control plane.
+
+### User Stories
+- As a Cluster Administrator I want to create one or more HostedClusterInfrastructure resources for my team to consume.
+  
+  This approach pre-creates the HostedClusterInfrastructure before the consumer creates the HostedCluster and NodePool resources. This approach has two advantage for ACM use cases:
```
Comment (enxebre) [inline enhancements/hypershift/hostedClusterInfrastructure.md:56]: nit: this says two advantages but then you list quite more

File: enhancements/hypershift/hostedClusterInfrastructure.md [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+## Glossary
+- Untangling HypershiftDeployment, HostedCluster and NodePool Google document - [Google Document - Option D](https://docs.google.com/document/d/1VabhUQa_uZlWO6mf9UHS5lGtnH_X915cih4ARJbNibc/edit#heading=h.iuee7tji8zz4)
+
+## Motivation
+ACM and MCE primary Cloud Provider use cases already provision these resources and manage these resources, this provides feature parity with Standalone flows that ACM Cluster Administrator's are used to.  This also allows to additional opporunities:
+  1. Infrastructure can be pre-created using an ACM Cloud Provider credential or by pre-populating the Custom Resource with values and applying it to the ACM/MCE hub. These can then be consumed by the ACM/MCE UI or CLI, and is a nice option, as the AWS credential does not need to be exposed to the user creating the HostedCluster.
+  2. You can re-use the HostedClusterInfrastructure resource between HostedClusters. This is especially useful in Azure, where it takes longer to create infrastructure then the hosted control plane.
+
+### User Stories
+- As a Cluster Administrator I want to create one or more HostedClusterInfrastructure resources for my team to consume.
```
Comment (enxebre) [inline enhancements/hypershift/hostedClusterInfrastructure.md:54]: I'd say this is a cluster service provider / infra admin, for a cluster admin we refer to the persona with full rbac in a guest cluster https://hypershift-docs.netlify.app/reference/concepts-and-personas/

File: enhancements/hypershift/hostedClusterInfrastructure.md (hunk 9 above)
Comment (jnpacker) [reply enhancements/hypershift/hostedClusterInfrastructure.md:63]: done.

File: enhancements/hypershift/hostedClusterInfrastructure.md (hunk 11 above)
Comment (jnpacker) [reply enhancements/hypershift/hostedClusterInfrastructure.md:56]: done

File: enhancements/hypershift/hostedClusterInfrastructure.md (hunk 12 above)
Comment (jnpacker) [reply enhancements/hypershift/hostedClusterInfrastructure.md:54]: I do think it ends up being a little confusing as Cluster Administrator is used earlier in the doc with MCE/ACM, but in our case that persona is the manager of multiple clusters (guess that definition will should be adjusted)

Comment (jnpacker) [reply enhancements/hypershift/hostedClusterInfrastructure.md]: I tried to make it more clear by including ACM/MCE and a bracket about infrastructure administrator.

**Comment by jnpacker** [comment]
It has been agreed with @enxebre and @zanetworker to put this on the shelf till 4.14, and then re-evaluate if we still think we need this.

**Pull request closed without merging**